**BISECTION METHOD**

Let's start with the exploration of the bisection method for root finding. 

This method takes the values of two extremes that we are sure contains one and only one root, finds the middle point and compares the value of the function calculated at the lower extreme with the function at the middle: if the two values so obtained are of opposite sign, i.e. if their product is negative, this means that the root is located between the middle point and the lower extreme and the middle point will become our new upper extreme. Otherwise, the root will be located between the middle point and the upper extreme and the middle point will become our new lower extreme. This operation is repeated until our condition for the loop is true, which is usually dependent on the difference between the upper and lower values. This guarantees that the root will be located in an interval of the chosen width. The simplicity of this method makes it a good fit for most functions, given that we know where the root we are interested in is approximately located, but is limited by our knowledge of the function and can also take a while to compute according to the extremes and the stopping point we choose. The number of steps needed can be predicted by calculating how many times the initial interval will have to be split in two until we get to the desired width. This can be calculated by taking: 

$$ floor \left (\log \dfrac{ |l-u|}{h} \right) + 1 $$

since the logarithm in base two of the value we usually obtain is not a whole number. 

The way it is computed could a problem in the unlikely case that our root is exactly in the middle of our interval. In this case the program will keep looking for the root in the upper half of the interval, since the product between the value of the function in the middle and the value in the lower extreme will not be negative. This will cause the method to converge to a value that is not a root. We can see an example of this with the function $f_1(x) = x^2 - 9$, looking for a root between $x = 2$ and $x = 4$.     

In [21]:
from math import floor, log

def bisection_step(f, bounds):
    """This function executes a step of the bisection method, taking as arguments the function to evaluate and a tuple containing
    upper and lower bounds in this order."""
    lower, upper = bounds  
    middle = (lower + upper)/2 
    if f(lower)*f(middle) < 0 : 
        return (lower, middle)
    else:
        return (middle, upper)


def f_1(x):
    return x**2 - 9

n, l, u = 0, 2, 4

p = floor(log(abs(l-u)/1e-8, 2)) + 1 #this is the formula I described above 

print("The predicted number of steps is", p)     

print("\n {:^15}  {:^15}  {:^15}  {:^15}".format("number of steps","lower", "upper", "difference"))

print("{:^15d}  {:15.12f}  {:15.12f}  {:15.12f}".format(n, l, u, u-l))

while abs(l - u) > 1e-8 and n < 1000:
    """This loop iterates the bisection method step as long as the distance between the two bounds is bigger than 10^-8, 
    displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = bisection_step(f_1, (l, u)) 
    n += 1
    print("{:^15d}  {:15.12f}  {:15.12f}  {:15.12f}".format(n, l, u, u-l)) 

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(abs(u+l)/2, f_1(abs(u+l)/2)))

The predicted number of steps is 28

 number of steps       lower            upper         difference   
       0          2.000000000000   4.000000000000   2.000000000000
       1          3.000000000000   4.000000000000   1.000000000000
       2          3.500000000000   4.000000000000   0.500000000000
       3          3.750000000000   4.000000000000   0.250000000000
       4          3.875000000000   4.000000000000   0.125000000000
       5          3.937500000000   4.000000000000   0.062500000000
       6          3.968750000000   4.000000000000   0.031250000000
       7          3.984375000000   4.000000000000   0.015625000000
       8          3.992187500000   4.000000000000   0.007812500000
       9          3.996093750000   4.000000000000   0.003906250000
      10          3.998046875000   4.000000000000   0.001953125000
      11          3.999023437500   4.000000000000   0.000976562500
      12          3.999511718750   4.000000000000   0.000488281250
      13          3.9997

To solve this problem, we just have to change our bisection formula step so that it takes the left interval if the product of the values of the function at the lower extremum and the value of the function at the middle of the interval is smaller **or equal** to zero.  

In [22]:
def bisection_step_corrected(f, bounds):
    """This function executes a step of the bisection method modified to include the case where the value of the function in the 
    middle of the interval is zero, taking as arguments the function to evaluate and a tuple containing upper and lower bounds 
    in this order."""
    lower, upper = bounds  
    middle = (lower + upper)/2 
    if f(lower)*f(middle) <= 0 : 
        return (lower, middle)
    else:
        return (middle, upper)

n, l, u = 0, 2, 4

p = floor(log(abs(l-u)/1e-8, 2)) + 1 

print("The predicted number of steps is", p)     

print("\n {:^15}  {:^15}  {:^15}  {:^15}".format("number of steps","lower", "upper", "difference"))

print("{:^15d}  {:15.12f}  {:15.12f}  {:15.12f}".format(n, l, u, u-l))

while abs(l - u) > 1e-8:
    """This loop iterates the bisection method step as long as the distance between the two bounds is bigger than 10^-8, 
    displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = bisection_step_corrected(f_1, (l, u)) 
    n += 1
    print("{:^15d}  {:15.12f}  {:15.12f}  {:15.12f}".format(n, l, u, u-l)) 

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format((u+l)/2, f_1((u+l)/2)))

The predicted number of steps is 28

 number of steps       lower            upper         difference   
       0          2.000000000000   4.000000000000   2.000000000000
       1          2.000000000000   3.000000000000   1.000000000000
       2          2.500000000000   3.000000000000   0.500000000000
       3          2.750000000000   3.000000000000   0.250000000000
       4          2.875000000000   3.000000000000   0.125000000000
       5          2.937500000000   3.000000000000   0.062500000000
       6          2.968750000000   3.000000000000   0.031250000000
       7          2.984375000000   3.000000000000   0.015625000000
       8          2.992187500000   3.000000000000   0.007812500000
       9          2.996093750000   3.000000000000   0.003906250000
      10          2.998046875000   3.000000000000   0.001953125000
      11          2.999023437500   3.000000000000   0.000976562500
      12          2.999511718750   3.000000000000   0.000488281250
      13          2.9997

This method works for any kind of function, given that we know an interval that only contains the root that we're interested in, and we can see this by looking for a root of the function $f_2(x) = x^8 + 2x^3 - 4$ between $x = 1$ and $x = 4$.  

In [23]:
def f_2(x):
    return x**8 + 2*x**3 - 4

n, l, u = 0, 1, 4

p = floor(log(abs(l-u)/1e-8, 2)) + 1 

print("The predicted number of steps is", p)     

print("\n {:^15}  {:^15}  {:^15}  {:^15}".format("number of steps","lower", "upper", "difference"))

print("{:^15d}  {:15.12f}  {:15.12f}  {:15.12f}".format(n, l, u, u-l))

while abs(l - u) > 1e-8:
    """This loop iterates the bisection method step as long as the distance between the two bounds is bigger than 10^-8, 
    displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = bisection_step_corrected(f_2, (l, u)) 
    n += 1
    print("{:^15d}  {:15.12f}  {:15.12f}  {:15.12f}".format(n, l, u, u-l)) 

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format((u+l)/2, f_2((u+l)/2)))

The predicted number of steps is 29

 number of steps       lower            upper         difference   
       0          1.000000000000   4.000000000000   3.000000000000
       1          1.000000000000   2.500000000000   1.500000000000
       2          1.000000000000   1.750000000000   0.750000000000
       3          1.000000000000   1.375000000000   0.375000000000
       4          1.000000000000   1.187500000000   0.187500000000
       5          1.000000000000   1.093750000000   0.093750000000
       6          1.046875000000   1.093750000000   0.046875000000
       7          1.046875000000   1.070312500000   0.023437500000
       8          1.058593750000   1.070312500000   0.011718750000
       9          1.058593750000   1.064453125000   0.005859375000
      10          1.058593750000   1.061523437500   0.002929687500
      11          1.060058593750   1.061523437500   0.001464843750
      12          1.060791015625   1.061523437500   0.000732421875
      13          1.0611

**REGULA FALSI METHOD**

Let's now explore a second method for root finding: the regula falsi. 
Here we approximate the function with a straight line that goes through the values of the function at the lower and at the upper bound, choosing as "middle point" the intercept of that straight line with the x axis. In other words, we are substituting the secant of that function through the two points. The values of the function at the lower bound and "middle point" are compared in the same way as in the previous method, returning a new couple of bounds at each step and iterating this with a loop. Using the same condition as before for this kind of loop would be unwise, since we are not halving the interval each time and there is no guarantee that the interval will actually be reduced in size when we get close to the desired root. We can see this if we look for again a root of the function $f_1(x) = x^2 - 9$ between $x = 1$ and $x = 4$.

In [24]:
def regula_falsi_step(f, bounds):
    """This function executes a step of the bisection method, taking as arguments the function to evaluate and a tuple containing
    upper and lower bounds in this order."""
    lower, upper = bounds
    m = (lower*f(upper) - upper*f(lower))/(f(upper) - f(lower))
    if f(lower)*f(m) < 0 :
        return (lower, m)
    else : 
        return (m, upper)

n, l, u = 0, 2, 4

print("\n {:^17}  {:^17}  {:^17}  {:^17}".format("number of steps","lower", "upper", "difference"))

print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l))

while abs(l - u) > 1e-8 and n < 1000:
    """This loop iterates the regula falsi method step as long as the distance between the two bounds is bigger than 10^-8, 
    displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = regula_falsi_step(f_1, (l, u)) 
    n += 1
    print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l)) 

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_1(l)))


  number of steps         lower              upper           difference    
        0          2.000000000000000  4.000000000000000  2.000000000000000
        1          2.833333333333333  4.000000000000000  1.166666666666667
        2          2.975609756097561  4.000000000000000  1.024390243902439
        3          2.996503496503496  4.000000000000000  1.003496503496504
        4          2.999500249875062  4.000000000000000  1.000499750124938
        5          2.999928602027703  4.000000000000000  1.000071397972297
        6          2.999989800185637  4.000000000000000  1.000010199814363
        7          2.999998542881539  4.000000000000000  1.000001457118461
        8          2.999999791840176  4.000000000000000  1.000000208159824
        9          2.999999970262882  4.000000000000000  1.000000029737118
       10          2.999999995751840  4.000000000000000  1.000000004248160
       11          2.999999999393120  4.000000000000000  1.000000000606880
       12          2.99

Even if the lower bound is exactly the value we were looking for, the loop continues to run since the difference between upper and lower bounds is still bigger than the value we have chosen. A better measure of the "closeness" to the exact value of the root is the value itself of the function, which should be close enough to zero in the chosen point to be satisfying. Let's see the same function with the new condition for the loop: 

In [5]:
n, l, u = 0, 2, 4 

print("\n {:^17}  {:^17}  {:^17}  {:^17}".format("number of steps","lower", "upper", "difference"))

print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l))

while abs(f_1(l)) > 1e-8 and abs(f_1(u)) > 1e-8:
    """This loop iterates the regula falsi method step as long as the value of the function at both the upper and lower bounds is
    bigger than 10^-8, displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = regula_falsi_step(f_1, (l, u)) 
    n += 1
    print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l)) 
    
print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_1(l)))


  number of steps         lower              upper           difference    
        0          2.000000000000000  4.000000000000000  2.000000000000000
        1          2.833333333333333  4.000000000000000  1.166666666666667
        2          2.975609756097561  4.000000000000000  1.024390243902439
        3          2.996503496503496  4.000000000000000  1.003496503496504
        4          2.999500249875062  4.000000000000000  1.000499750124938
        5          2.999928602027703  4.000000000000000  1.000071397972297
        6          2.999989800185637  4.000000000000000  1.000010199814363
        7          2.999998542881539  4.000000000000000  1.000001457118461
        8          2.999999791840176  4.000000000000000  1.000000208159824
        9          2.999999970262882  4.000000000000000  1.000000029737118
       10          2.999999995751840  4.000000000000000  1.000000004248160
       11          2.999999999393120  4.000000000000000  1.000000000606880

 The root is located a

This also shows that the new method is quicker than the bisection one, since it takes a smaller number of steps for the same function. It also solves the problem that brought us to modify the bisection step function, since if one of the extremes is close enough to the actual root it will be identified immediately. This situation will change quite rapidly the less close the function examined is to linear behaviour. Let's take for example $f_2(x) = x^8 + 2x^3 - 4$, which between $x = 1$ and $x = 4$ took 29 steps with the bisection method. 

In [6]:
In this report, we are analyzing data from neutron diffraction data collected at ISIS, the UK’s national neutron source, since it is a kind of radiation sensitive to magnetism. The material used is methylammonium cobalt(II) formate (CH3NH3Co(HCO2)3).n, l, u = 0, 1, 2

print("\n {:^17}  {:^17}  {:^17}  {:^17}".format("number of steps","lower", "upper", "difference"))

print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l))

while abs(f_2(l)) > 1e-8 and abs(f_2(u)) > 1e-8:
    """This loop iterates the regula falsi method step as long as the value of the function at both the upper and lower bounds is
    bigger than 10^-8, displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = regula_falsi_step(f_2, (l, u)) 
    n += 1
    print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l)) 

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_2(l)))


  number of steps         lower              upper           difference    
        0          1.000000000000000  2.000000000000000  1.000000000000000
        1          1.003717472118959  2.000000000000000  0.996282527881041
        2          1.007227303414260  2.000000000000000  0.992772696585740
        3          1.010539116534168  2.000000000000000  0.989460883465832
        4          1.013662300100489  2.000000000000000  0.986337699899511
        5          1.016605991121609  2.000000000000000  0.983394008878391
        6          1.019379060278457  2.000000000000000  0.980620939721543
        7          1.021990099973159  2.000000000000000  0.978009900026841
        8          1.024447415001483  2.000000000000000  0.975552584998517
        9          1.026759015688094  2.000000000000000  0.973240984311906
       10          1.028932613307821  2.000000000000000  0.971067386692179
       11          1.030975617605798  2.000000000000000  0.969024382394202
       12          1.03

Clearly, with 226 steps this method is not very efficient for this kind of function. What about transcendental functions like exponential and sine? As an example, I've taken the function $f_3(x) = exp(x) - 1$ between $x = -1$ and $x = 1$, and the function $f_4(x) = sin(x) - 1/2$ between $ x = \dfrac {\pi}{2} $  and $x = \pi$. 

In [7]:
from math import exp

def f_3(x):
    return exp(x) - 1

n, l, u = 0, -1, 1

print("\n {:^17}  {:^17}  {:^17}  {:^17}".format("number of steps","lower", "upper", "difference"))

print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l))

while abs(f_3(l)) > 1e-8 and abs(f_3(u)) > 1e-8:
    """This loop iterates the regula falsi method step as long as the value of the function at both the upper and lower bounds is
    bigger than 10^-8, displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = regula_falsi_step(f_3, (l, u)) 
    n += 1
    print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l)) 
    
print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_3(l)))


  number of steps         lower              upper           difference    
        0          -1.000000000000000  1.000000000000000  2.000000000000000
        1          -0.462117157260010  1.000000000000000  1.462117157260010
        2          -0.203030831979271  1.000000000000000  1.203030831979271
        3          -0.086811129005843  1.000000000000000  1.086811129005843
        4          -0.036646538125036  1.000000000000000  1.036646538125036
        5          -0.015383022923406  1.000000000000000  1.015383022923406
        6          -0.006441740127726  1.000000000000000  1.006441740127726
        7          -0.002694776300347  1.000000000000000  1.002694776300347
        8          -0.001126825652959  1.000000000000000  1.001126825652959
        9          -0.000471099943012  1.000000000000000  1.000471099943012
       10          -0.000196941337446  1.000000000000000  1.000196941337446
       11          -0.000082327916827  1.000000000000000  1.000082327916827
       12  

In [8]:
from math import sin, pi

def f_4(x):
    return sin(x) - 0.5

n, l, u = 0, pi/2, pi

print("\n {:^17}  {:^17}  {:^17}  {:^17}".format("number of steps","lower", "upper", "difference"))

print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l))

while abs(f_4(l)) > 1e-8 and abs(f_4(u)) > 1e-8:
    """This loop iterates the regula falsi method step as long as the value of the function at both the upper and lower bounds is
    bigger than 10^-8, displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = regula_falsi_step(f_4, (l, u)) 
    n += 1
    print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l)) 
    
print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_4(l)))


  number of steps         lower              upper           difference    
        0          1.570796326794897  3.141592653589793  1.570796326794897
        1          2.356194490192345  3.141592653589793  0.785398163397448
        2          2.586232286319997  3.141592653589793  0.555360367269796
        3          2.614934664598647  3.141592653589793  0.526657988991146
        4          2.617708122598263  3.141592653589793  0.523884530991530
        5          2.617967265862889  3.141592653589793  0.523625387726904
        6          2.617991400322655  3.141592653589793  0.523601253267138
        7          2.617993647319121  3.141592653589793  0.523599006270672
        8          2.617993856515818  3.141592653589793  0.523598797073975
        9          2.617993875992102  3.141592653589793  0.523598777597691

 The root is located at x = 2.617993876, where the function has value 1.73e-09


The method doesn't appear to have particular difficulties with solving equations containing transcendental functions, since when we are in a neighbourhood of such a function is closely approximated by the first couple of terms of its Taylor expansion, which makes it close enough to the behaviour of a linear function.

**NEWTON-RAPHSON METHOD**

The third and last method for root finding we are going to explore is the Newton-Raphson. Each steps uses an estimate of the root, the value of the function and the value of the derivative of the function at the estimate to find a more accurate estimate. The way this works is by subtracting the ratio of the two values of the function and of the derivative at the estimate from the given estimate, that is we move of an amount proportional to the value at the estimate in the direction opposite to the rate of change of the function. This means that we are moving in the direction where the function increases if we are at a negative value and in the direction where the function decreases if we are at a positive value, thus we are moving towards the zero! Since we have only one value and not two bounds, I will use the closeness to zero of the absolute value of the function as the stopping point of the loop that iterates a step of the method. To focus on the method alone and not being influenced by the method used to find the derivative, I am using the textbook derivatives for these functions.  

This method presents a quicker rate of convergence when compared to the previous ones, as we can see by evaluating once more the functions $f_1(x) = x^2 - 9$ close to $ x = 2$ and $f_2(x) =  x^8 + 2x^3 - 4$ close to $x = 1$. 

In [9]:
def newton_step(f, fp, x0):
    """"This function executes one step of the Newton-Rhapson method, taking the estimate, the function and the 
    derivative of the function as arguments. It returns the difference between the estimate and the ratio of the values
    of the function at the estimate and the derivative of the function at the estimate."""
    return x0 - f(x0)/fp(x0)

def fp_1(x):
    return 2*x
    
n, l = 0, 2    
    
print("\n {:^15}  {:^15}".format("number of steps", "estimate"))

print("{:^15d}  {:15.12f}".format(n, l))

while abs(f_1(l)) > 1e-8:
    """This loop iterates the Newton-Raphson method step as long as the value of the function at the estimate is bigger 
    than 10^-8, displaying the number of step executed and the value of the estimate for each step.""" 
    l = newton_step(f_1, fp_1, l)
    n += 1
    print("{:^15d}  {:15.12f}".format(n, l))

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_1(l)))


 number of steps     estimate    
       0          2.000000000000
       1          3.250000000000
       2          3.009615384615
       3          3.000015360039
       4          3.000000000039

 The root is located at x = 3.000000000, where the function has value 2.36e-10


In [10]:
def fp_2(x):
    return 8*x**7 + 6*x**2
    
n, l = 0, 1    
    
print("\n {:^15}  {:^15}".format("number of steps", "estimate"))

print("{:^15d}  {:15.12f}".format(n, l))

while abs(f_2(l)) > 1e-8:
    """This loop iterates the Newton-Raphson method step as long as the value of the function at the estimate is bigger 
    than 10^-8, displaying the number of step executed and the value of the estimate for each step.""" 
    l = newton_step(f_2, fp_2, l)
    n += 1
    print("{:^15d}  {:15.12f}".format(n, l))

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_2(l)))


 number of steps     estimate    
       0          1.000000000000
       1          1.071428571429
       2          1.061529756456
       3          1.061281096762
       4          1.061280944832

 The root is located at x = 1.061280945, where the function has value 1.07e-12


Efficiency when compared to the previous methods is clear-cut, as we can see from the table below: 

**Table of comparison of the three methods**

$f_1(x) = x^2 - 9$ with lower bound = 2 and upper bound = 4 

| Method used| Number of iterations |  Root Found  | Magnitude of the value of the function |
|----|----|----|----|----|
| Bisection   | 28  | 2.999999992549 | 1e-08 |
| Regula Falsi |  11 | 2.999999999393| 1e-09 |
| Newton-Raphson  | 4  |  3.000000000039 | 1e-10 |


$f_2(x) =  x^8 + 2x^3 - 4$ with lower bound = 1 and upper bound = 2

| Method used| Number of iterations |  Root Found  | Magnitude of the value of the function |
|----|----|----|----|----|
| Bisection   | 29  | 1.061280949041 | 1e-08 |
| Regula Falsi |  226 | 1.061280944315 | 1e-09 |
| Newton-Raphson  | 4  | 1.061280944832 | 1e-12 |


What could happen with different functions, though, is that the Newton-Raphson method returns a value that is not accurate.  

An example of this is provided by the function $f_5(x) = \exp (x) - 10^{-10}$, due to the fact that it converges very slowly so that after a certain point even if the function doesn't change significantly enough, the estimate point is still changing quite significantly. 

In [11]:
def f_5(x):
    return exp(x) - 0.0000000001

def fp_5(x):
    return exp(x)

n, l = 0, 1
    
print("\n {:^15}  {:^15}".format("number of steps", "estimate"))

print("{:^15d}  {:15.12f}".format(n, l))

while abs(f_5(l)) > 1e-8:
    """This loop iterates theNewton-Raphson method step as long as the value of the function at the estimate is bigger 
    than 10^-8, displaying the number of step executed and the value of the estimate for each step.""" 
    l = newton_step(f_5, fp_5, l)
    n += 1
    print("{:^15d}  {:15.12f}".format(n, l))

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_5(l)))

print("\n The correct value of the root is", log(0.0000000001))


 number of steps     estimate    
       0          1.000000000000
       1          0.000000000037
       2         -0.999999999863
       3         -1.999999999591
       4         -2.999999998852
       5         -3.999999996844
       6         -4.999999991384
       7         -5.999999976543
       8         -6.999999936200
       9         -7.999999826537
      10         -8.999999528441
      11         -9.999998718133
      12         -10.999996515489
      13         -11.999990528096
      14         -12.999974252771
      15         -13.999930012571
      16         -14.999809760559
      17         -15.999482921005
      18         -16.998594769316
      19         -17.996182665985
      20         -18.989641685896

 The root is located at x = -18.989641686, where the function has value 5.56e-09

 The correct value of the root is -23.025850929940457


To solve this problem, we just have to change the condition for stopping the iteration fo the step from one linked to the value of the function to one linked to how much the function has changed from the previous step. Once we've modified this, the algorithm functions as intended. 

In [12]:
n, l, u = 0, 1, 100    
    
print("\n {:^15}  {:^15}".format("number of steps", "estimate"))

print("{:^15d}  {:15.12f}".format(n, l))

while abs(l - u) > 1e-8:
    """This loop iterates the Newton-Raphson method step as long as the value of the estimate is changing 
    of an amount more than 10^-8, displaying the number of step executed and the value of the estimate for 
    each step.""" 
    u = l
    l = newton_step(f_5, fp_5, l)
    n += 1
    print("{:^15d}  {:15.12f}".format(n, l))

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_5(l)))

print("\n The correct value of the root is", log(0.0000000001))


 number of steps     estimate    
       0          1.000000000000
       1          0.000000000037
       2         -0.999999999863
       3         -1.999999999591
       4         -2.999999998852
       5         -3.999999996844
       6         -4.999999991384
       7         -5.999999976543
       8         -6.999999936200
       9         -7.999999826537
      10         -8.999999528441
      11         -9.999998718133
      12         -10.999996515489
      13         -11.999990528096
      14         -12.999974252771
      15         -13.999930012571
      16         -14.999809760559
      17         -15.999482921005
      18         -16.998594769316
      19         -17.996182665985
      20         -18.989641685896
      21         -19.971977379161
      22         -20.924801547130
      23         -21.802473555147
      24         -22.508238809246
      25         -22.912296921644
      26         -23.019640937364
      27         -23.025831687788
      28         -23.0258

This method can also present a considerably longer number of steps if we start our search very close to zero. In this case the ratio between the values of the function and the derivative will be quite small, thus the algorithm will move quite slowly in the direction of the actual root. We can see an example of this with the function $f_1(x) = x^2 - 9$, starting our search around several negative powers of $10$ and comparing the number of steps.  

In [13]:
for i in range(17):
    """This loop uses the Newton-Raphson method to find a root for different starting points, each time going
    a power of ten closer to zero. It then prints the starting point and the total number of steps executed before 
    the root was found."""
    l = 10**(-i)
    n = 0
    u = 100
    while abs(l - u) > 1e-8:
        """This loop iterates the Newton-Raphson method step as long as the value of the estimate is changing 
        of an amount more than 10^-8.""" 
        u = l
        l = newton_step(f_1, fp_1, l)
        n += 1  
    print("\n Starting from {:5.2e}, we find the root after {:^d} steps".format(10**(-i), n))


 Starting from 1.00e+00, we find the root after 6 steps

 Starting from 1.00e-01, we find the root after 10 steps

 Starting from 1.00e-02, we find the root after 13 steps

 Starting from 1.00e-03, we find the root after 16 steps

 Starting from 1.00e-04, we find the root after 20 steps

 Starting from 1.00e-05, we find the root after 23 steps

 Starting from 1.00e-06, we find the root after 26 steps

 Starting from 1.00e-07, we find the root after 30 steps

 Starting from 1.00e-08, we find the root after 33 steps

 Starting from 1.00e-09, we find the root after 36 steps

 Starting from 1.00e-10, we find the root after 40 steps

 Starting from 1.00e-11, we find the root after 43 steps

 Starting from 1.00e-12, we find the root after 46 steps

 Starting from 1.00e-13, we find the root after 50 steps

 Starting from 1.00e-14, we find the root after 53 steps

 Starting from 1.00e-15, we find the root after 56 steps

 Starting from 1.00e-16, we find the root after 60 steps


As we can see, the number of steps increases each time we get closer to zero, something that didn't happen with the other methods.  

**The black sheep of root finding: the asymptote**

A problem that is common to all root finding methods is to find a root that is close to an asymptote, which can either cause a non convergence, as in the case of the regula falsi method, or a convergence to a value that is not a root, as in the case of the bisection method. 

We can see this by looking for a root of the function $f_6 (x) = \dfrac {x - 5}{x - 4}$ between $x = 3$ and $x = 6$. 

In [14]:
def f_6(x):
    return (x-5)/(x-4)

n, l, u = 0, 3, 6 

print("\n {:^17}  {:^17}  {:^17}  {:^17}".format("number of steps","lower", "upper", "difference"))

print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l))

while abs(f_6(l)) > 1e-8 and abs(f_6(u)) > 1e-8 and n < 1000:
    """This loop iterates the regula falsi method step as long as the value of the function at both the upper and lower bounds is
    bigger than 10^-8, displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = regula_falsi_step(f_6, (l, u)) 
    n += 1
    print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l)) 

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_6(l))) 


  number of steps         lower              upper           difference    
        0          3.000000000000000  6.000000000000000  3.000000000000000
        1          7.000000000000000  6.000000000000000  -1.000000000000000
        2          3.000000000000001  6.000000000000000  2.999999999999999
        3          6.999999999999999  6.000000000000000  -0.999999999999999
        4          3.000000000000000  6.000000000000000  3.000000000000000
        5          7.000000000000000  6.000000000000000  -1.000000000000000
        6          3.000000000000001  6.000000000000000  2.999999999999999
        7          6.999999999999999  6.000000000000000  -0.999999999999999
        8          3.000000000000000  6.000000000000000  3.000000000000000
        9          7.000000000000000  6.000000000000000  -1.000000000000000
       10          3.000000000000001  6.000000000000000  2.999999999999999
       11          6.999999999999999  6.000000000000000  -0.999999999999999
       12        

       889         7.000000000000000  6.000000000000000  -1.000000000000000
       890         3.000000000000001  6.000000000000000  2.999999999999999
       891         6.999999999999999  6.000000000000000  -0.999999999999999
       892         3.000000000000000  6.000000000000000  3.000000000000000
       893         7.000000000000000  6.000000000000000  -1.000000000000000
       894         3.000000000000001  6.000000000000000  2.999999999999999
       895         6.999999999999999  6.000000000000000  -0.999999999999999
       896         3.000000000000000  6.000000000000000  3.000000000000000
       897         7.000000000000000  6.000000000000000  -1.000000000000000
       898         3.000000000000001  6.000000000000000  2.999999999999999
       899         6.999999999999999  6.000000000000000  -0.999999999999999
       900         3.000000000000000  6.000000000000000  3.000000000000000
       901         7.000000000000000  6.000000000000000  -1.000000000000000
       902        

The regula falsi has trouble with this because any function in the neighbourhood of one of its asymptotes has a behaviour similar to a high order polynomial.

In [15]:
n, l, u = 0, 3, 6

p = floor(log(abs(l-u)/1e-8, 2)) + 1 

print("The predicted number of steps is", p)     

print("\n {:^15}  {:^15}  {:^15}  {:^15}".format("number of steps","lower", "upper", "difference"))

print("{:^15d}  {:15.12f}  {:15.12f}  {:15.12f}".format(n, l, u, u-l))

while abs(l - u) > 1e-8:
    """This loop iterates the bisection method step as long as the distance between the two bounds is bigger than 10^-8, 
    displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = bisection_step_corrected(f_6, (l, u)) 
    n += 1
    print("{:^15d}  {:15.12f}  {:15.12f}  {:15.12f}".format(n, l, u, u-l)) 
    
print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format((l+u)/2, f_6((l+u)/2))) 

The predicted number of steps is 29

 number of steps       lower            upper         difference   
       0          3.000000000000   6.000000000000   3.000000000000
       1          3.000000000000   4.500000000000   1.500000000000
       2          3.750000000000   4.500000000000   0.750000000000
       3          3.750000000000   4.125000000000   0.375000000000
       4          3.937500000000   4.125000000000   0.187500000000
       5          3.937500000000   4.031250000000   0.093750000000
       6          3.984375000000   4.031250000000   0.046875000000
       7          3.984375000000   4.007812500000   0.023437500000
       8          3.996093750000   4.007812500000   0.011718750000
       9          3.996093750000   4.001953125000   0.005859375000
      10          3.999023437500   4.001953125000   0.002929687500
      11          3.999023437500   4.000488281250   0.001464843750
      12          3.999755859375   4.000488281250   0.000732421875
      13          3.9997

The bisection method has trouble with this because the function assumes values of opposite sign at the left of the asymptote when compared to the right, exactly like a root.  

In [16]:
def fp_6(x):
    return 1/(x-4)**2
    
n, l, u = 0, 3, 100    
    
print("\n {:^15}  {:^15}".format("number of steps", "estimate"))

print("{:^15d}  {:15.12f}".format(n, l))

while abs(l-u) > 1e-8 and n < 1000:
    """This loop iterates the Newton-Raphson method step as long as the value of the estimate is 
    changing of an amount bigger than 10^-8, displaying the number of step executed and the 
    value of the estimate for each step.""" 
    u = l
    l = newton_step(f_6, fp_6, l)
    n += 1
    print("{:^15d}  {:15.12f}".format(n, l))

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_6(l)))


 number of steps     estimate    
       0          3.000000000000
       1          1.000000000000
       2         -11.000000000000
       3         -251.000000000000
       4         -65530.999999999993
       5         -4294967290.999999046326
       6         -18446744073709543424.000000000000
       7         -340282366920938161231919703774474534912.000000000000
       8         -115792089237315989735501319857932638482122164996827441197474379810430211063808.000000000000
       9         -13407807929942549465471389561312667086994292072327229977483347439645365422331853474425488498520600322433332280977453460721137894551709517822781316115988480.000000000000
      10         -179769313486230313435132416858223013283261354317548374104623402464906114175265803317097681185405433648022198593022696178492809485617460480132534858739531251307816991132756480472671530906413122895311358559773208870689834174231107149755221833183778348295978703936325493682566845454613065195398408909457870290944.000

OverflowError: (34, 'Result too large')

In [17]:
n, l, u = 0, 6, 100    
    
print("\n {:^15}  {:^15}".format("number of steps", "estimate"))

print("{:^15d}  {:15.12f}".format(n, l))

while abs(l-u) > 1e-8 and n < 1000:
    """This loop iterates the Newton-Raphson method step as long as the value of the function at the estimate is bigger 
    than 10^-8, displaying the number of step executed and the value of the estimate for each step.""" 
    u = l
    l = newton_step(f_6, fp_6, l)
    n += 1
    print("{:^15d}  {:15.12f}".format(n, l))

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_6(l)))


 number of steps     estimate    
       0          6.000000000000
       1          4.000000000000


ZeroDivisionError: float division by zero

With this function, the Newton-Raphson method can either incur in an Overflow Error, since the derivative of the function in the vicinity of the asymptote becomes very large in absolute value, or in a division by zero error, since the derivative of the function at the asymptote is zero. This means that we will notice if the method has failed to return any value, let alone the correct value. 

**Machine precision error** 

If we then want our width to be of the order of $10^{-16}$ or smaller, we incur in a different kind of problem due to the fact that we are representing our numbers as floating points. Due to the fact that the maximum number of digits available with this representation is 17, to ask for such a precision would present us with all kinds of problems, including but not limited to the fact that we are setting the condition that two numbers are exactly equal from the point of view of the machine. We can see this with the function $f_1(x) = x^2 - 9$, looking for a root between $x = 2$ and $x = 4$: 

In [18]:
n, l, u = 0, 1, 3

p = floor(log(abs(l-u)/1e-17, 2)) + 1

print("The predicted number of steps is", p)     

print("\n {:^20}  {:^20}  {:^20}  {:^20}".format("number of steps","lower", "upper", "difference"))

print("{:^20d}  {:20.17f}  {:20.17f}  {:20.17f}".format(n, l, u, u-l))

while abs(l - u) > 1e-17 and n < 1000:
    """This loop iterates the bisection method step as long as the distance between the two bounds is bigger than 10^-17, 
    displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = bisection_step_corrected(f_1, (l, u)) 
    n += 1
    print("{:^20d}  {:20.17f}  {:20.18f}  {:20.18f}".format(n, l, u, u-l)) 
    
print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format((u+l)/2, f_1((u+l)/2)))

The predicted number of steps is 58

   number of steps            lower                 upper               difference     
         0             1.00000000000000000   3.00000000000000000   2.00000000000000000
         1             2.00000000000000000  3.000000000000000000  1.000000000000000000
         2             2.50000000000000000  3.000000000000000000  0.500000000000000000
         3             2.75000000000000000  3.000000000000000000  0.250000000000000000
         4             2.87500000000000000  3.000000000000000000  0.125000000000000000
         5             2.93750000000000000  3.000000000000000000  0.062500000000000000
         6             2.96875000000000000  3.000000000000000000  0.031250000000000000
         7             2.98437500000000000  3.000000000000000000  0.015625000000000000
         8             2.99218750000000000  3.000000000000000000  0.007812500000000000
         9             2.99609375000000000  3.000000000000000000  0.003906250000000000
     

Had I not put a limit to the number of steps, the loop would have tried to exercute an infinite number of steps, incurring in memory overflow. From the $52^{nd}$ step the function in the loop is trying to calculate a number that is between 2.99999999999999956 and 3.000000000000000000, which with machine precision is equal to 3.000000000000000000 and therefore the loop continues running with the same bounds. This is what happens for the Bisection method.  

In [19]:
n, l, u = 0, 2, 4 

print("\n {:^17}  {:^17}  {:^17}  {:^17}".format("number of steps","lower", "upper", "difference"))

print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l))

while abs(f_1(l)) > 1e-17 and abs(f_1(u)) > 1e-17 and n < 1000:
    """This loop iterates the regula falsi method step as long as the value of the function at both the upper and lower bounds is
    bigger than 10^-17, displaying the number of step executed, the upper and lower bounds and the difference for each step."""
    l, u = regula_falsi_step(f_1, (l, u)) 
    n += 1
    print("{:^17d}  {:17.15f}  {:17.15f}  {:17.15f}".format(n, l, u, u-l)) 
    
print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_1(l)))


  number of steps         lower              upper           difference    
        0          2.000000000000000  4.000000000000000  2.000000000000000
        1          2.833333333333333  4.000000000000000  1.166666666666667
        2          2.975609756097561  4.000000000000000  1.024390243902439
        3          2.996503496503496  4.000000000000000  1.003496503496504
        4          2.999500249875062  4.000000000000000  1.000499750124938
        5          2.999928602027703  4.000000000000000  1.000071397972297
        6          2.999989800185637  4.000000000000000  1.000010199814363
        7          2.999998542881539  4.000000000000000  1.000001457118461
        8          2.999999791840176  4.000000000000000  1.000000208159824
        9          2.999999970262882  4.000000000000000  1.000000029737118
       10          2.999999995751840  4.000000000000000  1.000000004248160
       11          2.999999999393120  4.000000000000000  1.000000000606880
       12          2.99

The value of the function after 18 steps is reasonably close to zero since it is of the order of $10^{-15}$, but since we have asked that the difference between zero and the value of the function at that point be of the order of $10^{-17}$ (impossible because of machine precision) the algorithm continues to run endlessly. This is what happens for the Regula Falsi method. 

In [20]:
n, l = 0, 2

print("\n {:^15}  {:^15}".format("number of steps", "estimate"))

print("{:^15d}  {:15.12f}".format(n, l))

while abs(l-u) > 1e-17 and n < 1000:
    """This loop iterates the Newton-Raphson method step as long as the value of the estimate is changing 
    of an amount larger than 10^-17, displaying the number of step executed and the value of the estimate 
    for each step.""" 
    u = l 
    l = newton_step(f_1, fp_1, l)
    n += 1
    print("{:^15d}  {:15.12f}".format(n, l))

print("\n The root is located at x = {:10.9f}, where the function has value {:5.2e}".format(l, f_1(l)))


 number of steps     estimate    
       0          2.000000000000
       1          3.250000000000
       2          3.009615384615
       3          3.000015360039
       4          3.000000000039
       5          3.000000000000
       6          3.000000000000

 The root is located at x = 3.000000000, where the function has value 0.00e+00


Using the Newton-Raphson method, we can have confirmation of the fact that the only way the loop will stop running is if the two values the loop condition taking the difference of are equal. With this method, it happens when the function at the estimate has value exactly zero, i.e. when the estimate coincides with the root with "infinite" precision. 

**CONCLUSION**

Each method has its strengths and weaknesses and, with the functions we are used to evaluate in the field of undegraduate physics, they don't usually present problems. If we wanted to combine these to find a new method that surpasses the three of them, we would have to: 

1) Plot the function we're interested in to visually estimate the approximate position of the root we want to find (if there is more than one root, we can iterate the checklist for each one)

2.1) If the root is not close to zero and not close enough to an asymptote that we can't discern between the two, use the Newton-Raphson method to find it

2.2) If the root is not close to zero but too close to an asymptote to discern between the two, use the Bisection method, find a point, print the value of the function at that point

- If the value of the function is very small, you have found the root

- If the value of the function is very large, you have found the asymptote and can now repeat step 2.2 with an interval that doesn't include it

- If the value of the function is very small but not satisfyingly accurate, we use the Newton-Raphson method to refine it